### PR TITLE
[clang][include-tree] Load even spurious modular dependencies

### PR DESCRIPTION
**Pull request by jansvoboda11**
With implicit modules, some modular dependencies can be "spurious". By that I mean that `#include "Header.h"` results in a PCM file load, but the header ends up being included textually anyways. (This can happen in situations that would trigger the `-Wincomplete-umbrella` warning.)

This patch ensures this behavior is preserved with include-tree. Previously, include-tree would simply textually include the header file and skip the PCM file load. This now ensures consistent behavior in edge-cases like the one in the attached test case (with `-fmodule-name=X`).

rdar://121220983

### CODE REVIEW

File: clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#include <Mod.h>
+//--- frameworks/Spurious.framework/Headers/Missing.h
+
+//--- include/module.modulemap
+module Mod { header "Mod.h" }
+//--- include/Mod.h
+typedef int mod_int;
+
+//--- tu.m
+#include <Spurious/Missing.h>
```
Comment (jansvoboda11) [inline clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c:40]: With implicit modules, this include will actually load the "Spurious" PCM and transitively "Mod" PCM. The scanner sees that this is unnecessary, since "Missing.h" is being treated textually anyways, and marks it as such in include-tree.

File: clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include <Mod.h>
```
Comment (jansvoboda11) [inline clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c:41]: Since the above include just textually includes "Missing.h", just marking Mod as visible here doesn't do anything. We need to either enter it textually, or import it properly (like we do with implicit modules and non-caching explicit modules).

**Comment by jansvoboda11** [comment]
Pushed new commit that approaches the fix from different angle. It introduces new include-tree node that attempts to mimic what happens in implicit builds and non-caching explicit builds: Spurious.pcm is loaded, the Missing submodule is found to be missing, and the header is included textually instead. Loading Spurious.pcm also loads Mod.pcm, so importing Mod for visibility only has the desired effect of making its symbols visible.

File: clang/include/clang/CAS/IncludeTree.h [hunk 3]
```diff
@@ -356,13 +358,50 @@ class IncludeTree::ModuleImport : public IncludeTreeBase<ModuleImport> {
...
+
+  Expected<IncludeTree> getIncludeTree() {
+    std::optional<ObjectProxy> Proxy;
+    if (llvm::Error Err = getCAS().getProxy(getReference(1)).moveInto(Proxy))
+      return Err;
+    return IncludeTree(*Proxy);
+  }
+
+  static bool isValid(ObjectStore &DB, ObjectRef Ref) {
+    // TODO: Implement this.
```
Comment (benlangmuir) [inline clang/include/clang/CAS/IncludeTree.h]: Should implement this TODO 😸

File: clang/lib/CAS/IncludeTree.cpp [hunk 4]
```diff
@@ -686,6 +694,12 @@ llvm::Error IncludeTree::Node::print(llvm::raw_ostream &OS, unsigned Indent) {
     return getIncludeTree().print(OS, Indent);
   case NodeKind::ModuleImport:
     return getModuleImport().print(OS, Indent);
+  case NodeKind::SpuriousImport: {
+    OS << "SpuriousImport\n";
+    if (auto Err = llvm::cantFail(getSpuriousImport().getModuleImport()).print(OS, Indent))
+      return Err;
+    return llvm::cantFail(getSpuriousImport().getIncludeTree()).print(OS, Indent);
```
Comment (benlangmuir) [inline clang/lib/CAS/IncludeTree.cpp]: Why are these `cantFail`? Typically we would just return the Error if there is one.

File: clang/lib/Frontend/FrontendAction.cpp [hunk 5]
```diff
@@ -997,8 +997,9 @@ bool FrontendAction::BeginSourceFile(CompilerInstance &CI,
   std::optional<cas::IncludeTreeRoot> IncludeTreeRoot;
   std::optional<StringRef> IncludeTreePCHBuffer;
   if (Input.isIncludeTree()) {
-    if (llvm::Error E = cas::IncludeTreeRoot::get(CI.getOrCreateObjectStore(),
-                                                  Input.getIncludeTree())
+    llvm::cas::ObjectStore &DB = CI.getOrCreateObjectStore();
+
+    if (llvm::Error E = cas::IncludeTreeRoot::get(DB, Input.getIncludeTree())
```
Comment (benlangmuir) [inline clang/lib/Frontend/FrontendAction.cpp]: Is this change needed?

File: clang/lib/Frontend/IncludeTreePPActions.cpp [hunk 6]
```diff
@@ -116,65 +116,85 @@ class IncludeTreePPActions final : public PPCachedActions {
...
+      return MakeModuleImport(Node->getModuleImport());
+    case cas::IncludeTree::NodeKind::Tree: {
+      auto IncludeTree = MakeIncludeTree(Node->getIncludeTree());
+      assert(std::holds_alternative<IncludeFile>(IncludeTree));
+      return std::get<IncludeFile>(IncludeTree);
+    }
+    case cas::IncludeTree::NodeKind::SpuriousImport:
+      auto SpuriousImportNode = Node->getSpuriousImport();
+      auto ModuleImport = MakeModuleImport(
+          llvm::cantFail(SpuriousImportNode.getModuleImport()));
```
Comment (benlangmuir) [inline clang/lib/Frontend/IncludeTreePPActions.cpp]: Why `cantFail` (same below)?  If there's a CAS error this will assert or be UB in a release build

**Review by benlangmuir — COMMENTED** [review]
I suggest we upstream the PPCallbacks change.  Relatedly, should anywhere you have `if (SuggestedModule && ModuleImported)` be simplified to `if (ModuleImported)` ? It seems like `SuggestedModule` should never be null if we're importing it, right?

File: clang/lib/Frontend/FrontendAction.cpp (hunk 5 above)
Comment (jansvoboda11) [reply clang/lib/Frontend/FrontendAction.cpp]: No, must be a leftover. I'll remove this.

**Comment by jansvoboda11** [comment]
> I suggest we upstream the PPCallbacks change.

Yes, if this approach gets a green light, I'll upstream this change.

> Relatedly, should anywhere you have `if (SuggestedModule && ModuleImported)` be simplified to `if (ModuleImported)` ? It seems like `SuggestedModule` should never be null if we're importing it, right?

That's correct. I'll update any current uses of `SuggestedModule && ModuleImported`.

**Comment by jansvoboda11** [comment]
Sorry, I should've mentioned this is more of a proof-of-concept that I wanted to get some high-level feedback on. I'm happy to implement the TODO and deal with errors properly if we agree this approach makes sense.

File: clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
+// REQUIRES: ondisk_cas
```
Comment (benlangmuir) [inline clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c:1]: I mentioned this to you offline, but putting here for posterity: we should extend this test to print the include-tree and check the structure of the new node. That helps clarify exactly what the test is doing.

File: clang/lib/Lex/PPDirectives.cpp [hunk 8]
```diff
@@ -2017,9 +2018,32 @@ void Preprocessor::HandleIncludeDirective(SourceLocation HashLoc,
         EnterAnnotationToken(SourceRange(HashLoc, EndLoc),
                              tok::annot_module_begin, SM);
       }
-      return;
-    }
-    if (auto *Import = std::get_if<PPCachedActions::IncludeModule>(&Include)) {
+    };
+
+    auto LoadModule = [&](const PPCachedActions::IncludeModule *Import) {
```
Comment (benlangmuir) [inline clang/lib/Lex/PPDirectives.cpp:2034]: Maybe factor `HandleIncludeMod` to call `LoadModule` and make the diagnostic part conditional? The error-handling logic is sufficiently interesting to not want it repeated.

File: clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+// RUN: %deps-to-rsp %t/cas-deps.json --module-name=Mod      > %t/cas-Mod.cc1.rsp
+// RUN: %deps-to-rsp %t/cas-deps.json --module-name=Spurious > %t/cas-Spurious.cc1.rsp
+// RUN: %deps-to-rsp %t/cas-deps.json --tu-index=0           > %t/cas-tu.rsp
+
+// RUN: cat %t/cas-tu.rsp | sed -E 's|.*"-fcas-include-tree" "(llvmcas://[[:xdigit:]]+)".*|\1|' > %t/tu.casid
+// RUN: clang-cas-test -cas %t/cas -print-include-tree @%t/tu.casid > %t/tu-include-tree.txt
+// RUN: FileCheck %s -input-file %t/tu-include-tree.txt -DPREFIX=%/t
+// CHECK:      [[PREFIX]]/tu.m llvmcas://
+// CHECK-NEXT: 1:1 <built-in> llvmcas://
+// CHECK-NEXT: 2:1 (Module) Spurious.Missing [[PREFIX]]/frameworks/Spurious.framework/Headers/Missing.h llvmcas://
```
Comment (benlangmuir) [inline clang/test/ClangScanDeps/modules-include-tree-implementation-transitive.c]: It seems like we should mark this as a spurious import explicitly in the output for readability.  Either inline or you could split it across lines and indent the import + tree.

**Review by benlangmuir — APPROVED** [review]
LGTM, just a couple of small suggestions
